src/lib.rs

```diff
@@ -0,0 +1,202 @@
+    let sites_term: Vector3<C64> = zip(magnitudes, eta.clone())
+        .map(|(m, e)| e * Complex::from(m))
+        .sum::<Vector3<C64>>();
+    let mut C = DMatrix::<C64>::zeros(n_sites, n_sites);
+    for c in &couplings {
+        *C.index_mut((c.index2, c.index2)) +=
+            (eta[c.index2].transpose() * c.matrix * sites_term).into_scalar();
+    }
+
+    q_vectors
+        .into_par_iter()
+        .map(|q| _spinwave_single_q(q, &C, n_sites, &z, &spin_coefficients, &couplings))
+        .collect()
+}
+
+/// Calculate spectra for a single q-value.
+#[allow(non_snake_case)]
+fn _spinwave_single_q(
+    q: Vector3<f64>,
+    C: &DMatrix<C64>,
+    n_sites: usize,
+    z: &[Vector3<C64>],
+    spin_coefficients: &DMatrix<C64>,
+    couplings: &Vec<&Coupling>,
+) -> Vec<C64> {
+    // create A and B matrices for the Hamiltonian
+
+    let mut A = DMatrix::<C64>::zeros(n_sites, n_sites);
+    let mut B = DMatrix::<C64>::zeros(n_sites, n_sites);
+
+    for c in couplings {
+        let phase_factor = ((2. * J * PI) * q.dot(&c.inter_site_vector)).exp();
+        let (i, j) = (c.index1, c.index2);
+
+        A[(i, j)] += (z[i].transpose() * c.matrix * z[j].conjugate()).into_scalar() * phase_factor;
+        B[(i, j)] += (z[i].transpose() * c.matrix * z[j]).into_scalar() * phase_factor;
+    }
+
+    A = A.component_mul(spin_coefficients);
+    B = B.component_mul(spin_coefficients);
+
+    // create Hamiltonian as a block matrix (the stack! macro creates a block matrix)
+    let A_minus_C: DMatrix<C64> = A.clone().sub(C);
+    let A_conj_minus_C: DMatrix<C64> = A.adjoint().sub(C);
+    let hamiltonian: DMatrix<C64> = stack![ A_minus_C, B; 
+                                            B.adjoint(), A_conj_minus_C];
+
+    // try to take square root of Hamiltonian with Cholesky; if it fails, use LDL
+    let sqrt_hamiltonian = {
+        let (l, d) = ldl(hamiltonian);
+        l * DMatrix::from_diagonal(&d.map(nalgebra::Complex::sqrt))
+        //match hamiltonian.cholesky() {
+        //  Some(m) => m.l(),
+        //None => panic!(), // hamiltonian.ldl().unwrap().l
+        //}
+    };
+
+    // 'shc' is "square root of Hamiltonian with commutation"
+    // We need to enforce the bosonic commutation properties, we do this
+    // by finding the 'square root' of the matrix (i.e. finding K such that KK^dagger = H)
+    // and then negating the second half.
+    //
+    // In matrix form we do
+    //
+    //     M = K^dagger g K
+    //
+    // where g is a diagonal matrix of length 2n, with the first n entries being 1, and the
+    // remaining entries being -1. We do this by just multiplying the >n_sites columns of shc.
+    let mut shc: DMatrix<C64> = sqrt_hamiltonian.clone();
+    let mut negative_half = shc.view_mut((0, n_sites), (2 * n_sites, n_sites));
+    negative_half *= Complex::from(-1.);
+
+    if let Some(v) = (shc.adjoint() * sqrt_hamiltonian).eigenvalues() {
+        v.data.into()
+    } else {
+        panic!()
```

Comment:
Can we add an error message here, like `Could not compute eigenvalues of the Hamiltonian`, please?